_drafts/profiling-firmware-on-cortex-m.md

```diff
@@ -0,0 +1,682 @@
+    % FUNCTION
+ 0.05 *SLEEP*
+51.94 mandel
+32.64 spi_xfer
+ 8.40 lcd_command
+ 4.63 lcd_draw_pixel
+ 2.17 msleep
+ 0.10 uart_putc
+ 0.02 _realloc_r
+ 0.01 rcc_clock_setup_pll
+ 0.01 sys_tick_handler
+ 0.00 __fputwc
+ 0.00 _fclose_r
+ 0.00 __sfvwrite_r
+ 0.00 gpio_mode_setup
+ 0.00 _dtoa_r
+-----
+ 100% 23170 samples
+```
+
+As you can see, this data is much more detailed than what we got with the Poor
+Man's Profiler approach. In fact, we can see that we are spending most of our
+time in the `mandel` function, and not in the `spi_xfer` function as we
+thought.
+
+> Note: In the process of writing this blog post, I stumbled upon the Orbuculum
+> project (https://github.com/orbcode/orbuculum). While this looks very
+> promising for ITM/ETM and other debugging, I was not able to get it to work on
+> MacOS.
+
+### Profiling Individual Functions
+
+So far, we have been profiling our entire program. This is useful to identify
+overall bottlenecks, but oftentimes we are concerned only with a specific code
+path. For example, we may want to profile a specific message parser, or an
+interrupt service routine.
+
+To do this, we will need to modify our firmware code a bit. Our goal is to write
+a `PROFILE` macro which can be used to enable ITM profiling for a specific set
+of code.
+
+In our Mandelbrot example, let's profile `lcd_show_frame`. We want our macro to
+be invoked as such:
+
+```c
+int main(void) {
+    ...
+    PROFILE(lcd_show_frame);
+    ...
+}
+```
+
+What does our macro need to do? At a high level:
+1. Enable PC sampling
+2. Execute the code being profiled
+3. Disable PC sampling
+
+We can start filling in our macro:
+
+```c
+#define PROFILE(code) {         \
+  // enable PC sampling         \
+  code                          \
+  // disable PC sampling        \
+}
+```
+
+To enable or disable PC sampling, we need to write to the DWT Control Register.
+Thankfully, libopencm3 has definition for this register[^libopencm3-dwt] and its fields which
+makes our life easy.
+
+To enable PC sampling:
+```c
+DWT_CTRL &= ~DWT_CTRL_POSTPRESET;
+DWT_CTRL |= 0x3 << DWT_CTRL_POSTPRESET_SHIFT;
+DWT_CTRL |= DWT_CTRL_CYCTAP;
+DWT_CTRL |= DWT_CTRL_PCSAMPLENA;
+```
+
+and to disable it:
+
+```c
+DWT_CTRL &= ~DWT_CTRL_POSTPRESET;
+DWT_CTRL &= ~DWT_CTRL_CYCTAP;
+DWT_CTRL &= ~DWT_CTRL_PCSAMPLENA;
+```
+
+We wrap this into a function to make our life easier:
+
+```c
+static void dwt_pcsampler_enable(bool enable) {
+    if (enable) {
+        DWT_CTRL &= ~DWT_CTRL_POSTPRESET;
+        DWT_CTRL |= 0x3 << DWT_CTRL_POSTPRESET_SHIFT;
+        DWT_CTRL |= DWT_CTRL_CYCTAP;
+        DWT_CTRL |= DWT_CTRL_PCSAMPLENA;
+    } else {
+        DWT_CTRL &= ~DWT_CTRL_POSTPRESET;
+        DWT_CTRL &= ~DWT_CTRL_CYCTAP;
+        DWT_CTRL &= ~DWT_CTRL_PCSAMPLENA;
+    }
+}
+```
+
+and fill in our macro:
+
+```
+#define PROFILE(code)                   \
+    {                                   \
+        dwt_pcsampler_enable(true);     \
+        code                            \
```

Comment:
is code a function call?

Should
```
int main(void) {
    ...
    PROFILE(lcd_show_frame);
    ...
}
```

be

```
int main(void) {
    ...
    PROFILE(lcd_show_frame());
    ...
}
```
?

Reply:
duh 🤦‍♂️